### DISCUSSION TITLE
Load encrypted model

### DISCUSSION THREAD

**Discussion by mjto** [deepjavalibrary/djl#716] [Ideas]
For the deployment of deep learning model, I have the need to encrypt the trained model that will be saved locally on end user computer. Then before running inference, the model will be loaded (by ModelZoo.loadModel) after decryption. Ideally, I'd like to keep the decrypted model in memory.

The above decryption and model loading can be achieved smoothly if ModelZoo.loadModel takes stream as input (the result of the decryption).  I can see that there exists low level function that does take stream as input such as torch::jit::load(std::istream &in, …) but is not wrapped and exposed in DJL. 

My question is whether such feature (ModelZoo.loadModel takes stream as input) or something similar is planned to be added in DJL in the near future. This seems something very useful for certain application scenarios but correct me if there is already very straightforward way to implement model encryption and decryption while using DJL. 

Many thanks.

**Comment by lanking520** [comment]
@mjto Recently when dealing with DJL parameter saving, I made some changes to allow the InputStream input for PyTorch and MXNet: https://github.com/awslabs/djl/commit/92cc44881a65c4209eaa4a9a79034803236f9ed7. However, we haven't implement the loading mechanism to load directly from InputStream, we can add one if needed.

**Reply by ming-jiang-tfs** [reply]
Thanks a lot! This is exactly what I am talking about.
It would be great if this (loading model from stream) can be implemented at ModelZoo level (ModelZoo.loadModel).

**Reply by lanking520** [reply]
I think we may add it specifically to PyTorch. The modelZoo model usually is very complicated, sometimes in zip format. Scanning in zipstream isn't very straight forward. probably something like

```
PtModel.load(InputStream)
```
should be doable. @frankfliu

**Reply by mjto** [reply]
Adding to PyTorch is good enough. It will solve most of my problem under discussion.
Very appreciate it.

**Reply by lanking520** [reply]
> Adding to PyTorch is good enough. It will solve most of my problem under discussion.
> 
> Very appreciate it.
> 
> 
> 
> 

Please file an issue with Feature request title. Will be happy to have that for you 😀

**Comment by lanking520** [comment]
@mjto https://github.com/awslabs/djl/pull/729

**Comment by lanking520** [comment]
Hi @mjto, Since the PR is merged. You should be able to try it out loading from PyTorch directly. Let me know if there is any more issues.

**Reply by mjto** [reply]
Thank you for getting this done quickly. I will try it out.
